File: 155MinStack.cpp
#include<iostream>
#include<stack>
using namespace std;

class MinStack {
public:
    MinStack() {
        
    }
    
    void push(int val) {
        stk.push(val);
        if (minStk.empty()) {
            minStk.push(val);
        } else if (val < minStk.top()) {
            minStk.push(val);
        } else if(val > minStk.top()) {
            minStk.push(minStk.top());
        }

    }
    
    void pop() {
        stk.pop();
        minStk.pop();
        
    }
    
    int top() {
        return stk.top();
        
    }
    
    int getMin() {
        return minStk.top();
        
    }
private:
    stack<int> stk;
    stack<int> minStk;

};

int main () {
    MinStack s;

}

File: 155MinStack.py
class MinStack(object):

    def __init__(self):
        self.stack = []
        self.minStack = []

        

    def push(self, val):
        self.stack.append(val)
        if len(self.minStack) == 0:
            self.minStack.append(val)
        else:
            if val < self.minStack[-1]:
                self.minStack.append(val)
            else:
                self.minStack.append(self.minStack[-1])
        

    def pop(self):
        self.stack.pop()
        self.minStack.pop()
        

    def top(self):
        return self.stack[-1]
        

    def getMin(self):
        return self.minStack[-1]
    

if __name__ == "__main__":
    s = MinStack()
    s.push(2)
    s.push(3)
    s.push(4)
    s.push(1)

    print(s.getMin())
    print(s.pop())
    print(s.getMin())

